### PR TITLE
Fix for github issue 34. The parsing logic is changed.

### PR DESCRIPTION
**Pull request by msenmurugan**

### CODE REVIEW

File: pkg/docker/parseimage.go [hunk 1]
```diff
@@ -26,23 +26,17 @@ import (
 	"regexp"
 )
 
-var prefix = regexp.MustCompile("docker-pullable://")
-var digestRegexp = regexp.MustCompile("@sha256:([a-zA-Z0-9]+)$")
+var digestRegexp = regexp.MustCompile("^docker-pullable://(.+)@sha256:([a-zA-Z0-9]+)$")
```
Comment (rrati) [inline pkg/docker/parseimage.go]: Will this work for all openshift images?  openshift stores some images in etcd and those won't have a docker-pullable prefix.

File: pkg/docker/parseimage_test.go [hunk 2]
```diff
@@ -35,38 +35,67 @@ func TestParseImageIDString(t *testing.T) {
...
+			sha:         "cb4983d8399a59bb5ee6e68b6177d878966a8fe41abe18a45c3b1d8809f1d043",
+			shouldPass:  true,
+		},
+		{
+			description: "missing prefix",
 			prefix:      "",
 			name:        "abc/def",
 			sha:         "cb4983d8399a59bb5ee6e68b6177d878966a8fe41abe18a45c3b1d8809f1d043",
-			shouldPass:  true,
+			shouldPass:  false,
```
Comment (rrati) [inline pkg/docker/parseimage_test.go]: I believe this change is wrong.  This test is there specifically to test openshift's images in etcd.  They won't have a docker-pullable prefix.

File: pkg/docker/parseimage_test.go [hunk 3]
```diff
@@ -35,38 +35,67 @@ func TestParseImageIDString(t *testing.T) {
...
+		{
+			description: "valid format with 2 directories",
 			prefix:      "docker-pullable://",
 			name:        "abc/def",
 			sha:         "cb4983d8399a59bb5ee6e68b6177d878966a8fe41abe18a45c3b1d8809f1d043",
 			shouldPass:  true,
 		},
 		{
-			description: "no prefix",
+			description: "valid format with 3 directories",
```
Comment (rrati) [inline pkg/docker/parseimage_test.go]: Is 3 directories somehow different than 2?

Comment (msenmurugan) [reply pkg/docker/parseimage_test.go]: No @rrati.

File: pkg/docker/parseimage_test.go (hunk 2 above)
Comment (msenmurugan) [reply pkg/docker/parseimage_test.go]: Fixed the issue, can you please review it again.

File: pkg/docker/parseimage.go (hunk 1 above)
Comment (msenmurugan) [reply pkg/docker/parseimage.go]: Fixed the issue. Please review the changes again

File: pkg/docker/parseimage.go [hunk 4]
```diff
@@ -24,25 +24,61 @@ package docker
...
+
+	name, digest, err = ParseDockerImageString(imageID)
+	if err == nil {
+		return name, digest, fmt.Errorf("Scanning of static/builder pod image %s is not supported!!", imageID)
+	} else {
+		return name, digest, fmt.Errorf("Unable to match dockerPullableRegexp regex <%s> to input <%s>", dockerPullableRegexp.String(), imageID)
+	}
+}
+
+func ParseImageString(imageID string) (string, string, error) {
```
Comment (rrati) [inline pkg/docker/parseimage.go]: lowercase this function.  It doesn't need to be exported

File: pkg/docker/parseimage.go [hunk 5]
```diff
@@ -24,25 +24,61 @@ package docker
...
+	match := imageRegexp.FindStringSubmatch(imageID)
+	if len(match) != 3 {
+		return "", "", fmt.Errorf("unable to match imageRegexp regex <%s> to input <%s>", imageRegexp.String(), imageID)
+	}
+	name := match[1]
+	digest := match[2]
+	return name, digest, nil
+}
+
+func ParseDockerImageString(imageID string) (string, string, error) {
```
Comment (rrati) [inline pkg/docker/parseimage.go]: lowercase this function name

File: pkg/docker/parseimage.go [hunk 6]
```diff
@@ -24,25 +24,61 @@ package docker
...
+	match := dockerRegexp.FindStringSubmatch(imageID)
+	if len(match) != 2 {
+		return "", "", fmt.Errorf("unable to match dockerRegexp regex <%s> to input <%s>", dockerRegexp.String(), imageID)
+	}
+	digest := match[1]
+	return "", digest, nil
+}
+
+func ParseDockerPullableImageString(imageID string) (string, string, error) {
```
Comment (rrati) [inline pkg/docker/parseimage.go]: lowercase this function name

File: pkg/docker/parseimage_test.go (hunk 3 above)
Comment (rrati) [reply pkg/docker/parseimage_test.go]: Then why add a test for 3 dirs if 2 is the same?

File: pkg/docker/parseimage.go [hunk 7]
```diff
@@ -24,25 +24,61 @@ package docker
...
-		return "", "", fmt.Errorf("unable to match digestRegexp regex <%s> to input <%s>", digestRegexp.String(), str)
+
+	name, digest, err = ParseImageString(imageID)
+	if err == nil {
+		return name, digest, err
+	}
+
+	name, digest, err = ParseDockerImageString(imageID)
+	if err == nil {
+		return name, digest, fmt.Errorf("Scanning of static/builder pod image %s is not supported!!", imageID)
```
Comment (rrati) [inline pkg/docker/parseimage.go]: Static pods is only one case where this might occur.  It might be the only one, but I'm not sure.  Better not to call out the specific failure case by name

File: pkg/docker/parseimage.go [hunk 8]
```diff
@@ -24,25 +24,61 @@ package docker
...
 //   docker-pullable://registry.kipp.blackducksoftware.com/blackducksoftware/hub-registration@sha256:cb4983d8399a59bb5ee6e68b6177d878966a8fe41abe18a45c3b1d8809f1d043
 func ParseImageIDString(imageID string) (string, string, error) {
-	str := imageID
-	match := prefix.FindStringIndex(str)
-	if len(match) > 0 && match[0] == 0 {
-		str = str[match[1]:]
+	name, digest, err := ParseDockerPullableImageString(imageID)
+	// Since the GO doesn't support for "don't start with string regex", had an ugly fix with HasPrefix in below line
+	if err == nil || strings.HasPrefix(imageID, "docker-pullable://") {
+		return name, digest, err
```
Comment (rrati) [inline pkg/docker/parseimage.go]: I'm not sure I get this case.  Either the function returns an err and thus fails, or it success and err == nil.  Why do you need to do a HasPrefix check?  If err == nil then it succeed by definition, and if err != nil that name and digest are empty.

File: pkg/docker/parseimage_test.go [hunk 9]
```diff
@@ -67,6 +88,7 @@ func TestParseImageIDString(t *testing.T) {
 	for _, tc := range testcases {
 		imageID := fmt.Sprintf("%s%s@sha256:%s", tc.prefix, tc.name, tc.sha)
 		name, sha, err := ParseImageIDString(imageID)
+		fmt.Printf("Test: %s, err: %s, imageID: %s, name: %s, sha: %s \n", tc.description, err, imageID, name, sha)
```
Comment (rrati) [inline pkg/docker/parseimage_test.go]: Remove.  SHouldn't have fmt.Printf in tests

File: pkg/docker/parseimage_test.go [hunk 10]
```diff
@@ -76,5 +98,15 @@ func TestParseImageIDString(t *testing.T) {
 		if sha != tc.sha && tc.shouldPass {
 			t.Errorf("[%s] sha is wrong.  Expected %s got %s", tc.description, tc.sha, sha)
 		}
+		if !tc.shouldPass && err == nil {
+			t.Errorf("The error should not be empty, description: %s , prefix: %s, name: %s, sha: %s", tc.description, tc.prefix, name, sha)
+		}
+	}
+
+	imageID := "docker://sha256:cb4983d8399a59bb5ee6e68b6177d878966a8fe41abe18a45c3b1d8809f1d043"
```
Comment (rrati) [inline pkg/docker/parseimage_test.go]: We should be able to fold this test into the other iterative tests by re-working things a little.  Move the static @sha256 bit into the sha field for each test then should be able to put this test as a usual test case.

File: pkg/docker/parseimage.go (hunk 8 above)
Comment (msenmurugan) [reply pkg/docker/parseimage.go]: For the missing image name test case, the dockerPullableRegexp will error out and the err will be != nil and hence the code will start executing the next imageregex. The imageregex will not provide any error to missing image name test case and it will return image name as docker-pullabe:// and sha as <<some value>>. Golang doesn't support "don't start with string regex". Hence we need to add a hack to stop terminating if the image id starts with docker-pullable://

File: pkg/docker/parseimage.go (hunk 7 above)
Comment (msenmurugan) [reply pkg/docker/parseimage.go]: Removed static/builder from error message

File: pkg/docker/parseimage.go (hunk 6 above)
Comment (msenmurugan) [reply pkg/docker/parseimage.go]: Done

File: pkg/docker/parseimage.go (hunk 5 above)
Comment (msenmurugan) [reply pkg/docker/parseimage.go]: Done

File: pkg/docker/parseimage.go (hunk 4 above)
Comment (msenmurugan) [reply pkg/docker/parseimage.go]: Done

**Comment by msenmurugan** [comment]
The different error messages will get printed based on the existing code. See for sample code in image_annotator.go

```
_, imageSha, err := docker.ParseImageIDString(osImage.DockerImageReference)
if err != nil {
   metrics.RecordError("image_annotator", "unable to parse openshift imageID")
   log.Errorf("unable to parse openshift imageID from image %s: %v", imageName, err)
   continue
}
```

File: pkg/docker/parseimage_test.go (hunk 3 above)
Comment (msenmurugan) [reply pkg/docker/parseimage_test.go]: Removed 3 dirs test case
